src/main/scala/pt/tecnico/dsi/openstack/keystone/services/RoleAssignment.scala

```diff
@@ -1,69 +1,156 @@
 package pt.tecnico.dsi.openstack.keystone.services
 
 import cats.effect.Sync
+import cats.syntax.functor._
 import cats.syntax.flatMap._
 import fs2.Stream
 import org.http4s.Method.{HEAD, PUT}
-import org.http4s.Status.{NotFound, Successful}
-import org.http4s.client.{Client, UnexpectedStatus}
+import org.http4s.client.Client
 import org.http4s.{Header, Query, Uri}
-import pt.tecnico.dsi.openstack.common.services.{CrudService, Service}
-import pt.tecnico.dsi.openstack.keystone.models.Role
+import pt.tecnico.dsi.openstack.common.services.Service
+import pt.tecnico.dsi.openstack.keystone.models.{Group, Role, User}
 
-trait RoleAssignment[F[_]] { this: CrudService[F, _, _, _] =>
-  object roles {
-    /**
-      * Role assignments for users on this context.
-      */
-    val users = new RoleAssignmentService(uri, "users", authToken)
-    /**
-      * Role assignments for groups on this context.
-      */
-    val groups = new RoleAssignmentService(uri, "groups", authToken)
-  }
-}
-
-class RoleAssignmentService[F[_]: Sync: Client](val uri: Uri, target: String, authToken: Header) extends Service[F](authToken) {
+class RoleAssignment[F[_]: Sync: Client](uri: Uri, authToken: Header) extends Service[F](authToken) { self =>
   import dsl._
-
+  
+  final class Assign(roleId: String) {
+    private def to(subjectType: String, id: String): F[Unit] =
+      client.expect(PUT(uri / subjectType / id / "roles" / roleId, authToken))
+    
+    def toUser(id: String): F[Unit] = to("users", id)
+    def toGroup(id: String): F[Unit] = to("groups", id)
+    
+    // Dotty union types here would be awesome
+    def to(user: User): F[Unit] = toUser(user.id)
+    def to(group: Group): F[Unit] = toGroup(group.id)
+  }
+  final class Unassign(roleId: String) {
+    private def from(subjectType: String, id: String): F[Unit] =
+      self.delete(uri / subjectType / id / "roles" / roleId)
+    def fromUser(id: String): F[Unit] = from("users", id)
+    def fromGroup(id: String): F[Unit] = from("groups", id)
+    
+    // Dotty union types here would be awesome
+    def from(user: User): F[Unit] = fromUser(user.id)
+    def from(group: Group): F[Unit] = fromGroup(group.id)
+  }
+  final class Is(roleId: String) {
+    private def assignedTo(subjectType: String, id: String): F[Boolean] =
+      HEAD(uri / subjectType / id / "roles" / roleId, authToken).flatMap(client.expectOption(_)(void)).map(_.isDefined)
+    
+    def assignedToUser(id: String): F[Boolean] = assignedTo("users", id)
+    def assignedToGroup(id: String): F[Boolean] = assignedTo("groups", id)
+    
+    // Dotty union types here would be awesome
```

Comment:
Why? You would have to discriminate between user and group to call the `assignedTo` with the correct argument.

Reply:
If you are using the domain classes (eg: User or group) you can invoke `assignedTo` which is overloaded to differentiate between them. But if you want to invoke `assignedTo` using Ids there is no way to differentiate between an User id and a group id, so we need the `assignedToUser` and `assignedToGroup` functions.